python-mini-projects/projects/Send_email_from_csv/Sending_mail.py

```diff
@@ -0,0 +1,52 @@
+import csv
+from email.message import EmailMessage
+import smtplib
+
+
+def get_credentials(filepath):
+    with open("credentials.txt", "r") as f:
+        email_address = f.readline()
+        email_pass = f.readline()
+    return (email_address, email_pass)
+
+
+def login(email_address, email_pass, s):
+    s.ehlo()
+    # start TLS for security
+    s.starttls()
+    s.ehlo()
+    # Authentication
+    s.login(email_address, email_pass)
+    print("login")
+
+
+def send_mail():
+    s = smtplib.SMTP("smtp.gmail.com", 587)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *s*. The resource is allocated by call *smtplib.SMTP*. The resource is referenced at lines 26, 43, 47. The resource is closed at line 47. There are other execution paths that do not contain closure statements, for example, when *csv.reader()* throws an exception. To prevent this resource leak, close *s* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).